docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,206 @@
+1. Starting migration
+    - Migration can only be started after a Shoot cluster has been successfully created so that the `status.seed` field in the `Shoot` resource has been set
+    - The `Shoot` resource's field `spec.seedName="new-seed"` is edited to hold the name of the _New Seed_ and reconciliation is automatically triggered
+    - The Garden Controller Manager checks if the equality between `spec.seedName` and `status.seed`, detects that they are different and triggers migration.
+1. The Garden Controller Manager waits for the _New Seed_ to be ready
+1. ShootState is synced from the _Host Seed_ cluster
+    - The Garden Controller Manager lists all deployed extension resources for the Shoot and copies the values in their `status.state` field to the `ShootState` resource
+    - If the _Host Seed's_ Apiserver is not available, then last successfull synced `ShotoState` is used
+1. Stop the Shoot's api server
+1. Scale Down the Shoot's control plane in the _Host Seed_. ETCD backup will be done automatically before the ETCD pod is scaled down.
+1. Resume normal reconciliation flow in the _New Seed_. Extension resources are annotated with `core.gardener.cloud/operation=restore` to instruct the extension controllers to reconstruct their state.
+
+
+### Leader Election and Control Plane Termination
+
+During migration "split brain" scenario must be avoided. This means that a Shoot's control plane in the _Host Seed_ must be scaled down before it is scaled up in the _New Seed_.
+
+**Note**: This section is still under discussion. The plan is to first implement the `ShootState` and modify the reconcilation flow and extensions accordingly. Additionally multiple scenarios need to be considered depending on the reachability to the Garden cluster from the Host Seed components, the Host Seed's api server from the Garden cluster and the Host Seed's api server from the controllers running on the seed. The initial implementation will only cover the case where everything is running.
+
+Extension controllers do not need leader election functionality because they only reconcile the extension resources if the reconcile operation annotation is specified in the resource. Since this is set only during a reconciliation triggered by the Garden Controller Manager it cannot happen during migration.
+
+For other controllers in the controlplane (e.g MCM, etcd backup restore, kube apiserver) leader election has to be implemented. We plan to introduce gardenlet soon therefore the garden cluster is expected to be reachable from all seeds. Initially the Shoot's `spec.seedName` can be used to determine which seed is the current leader of the shoot.
+Another flag might be needed to tell the gardenlet in the _New Seed_ that the control plane in the _Host Seed_ has been scaled down and the Shoot reconciliation can begin.
+
+Horizontal Pod Autoscalers also need to be considered and removed.
+
+When discussing this section consider what happens in kubernetes:
```

Comment:
Yes, but here it's actually very poorly implemented. @amshuman-kr has therefore implemented a circuit-breaker in the shoot control plane that stops the cluster from falling apart if the kubelets can's reach the API server anymore, because otherwise, an even/otherwise healthy cluster falls apart within seconds with a major downtime for all workload. E.g. when the node becomes `NotReady`, it gets deregistered from the LBs and traffic ceases. Also, pod state is not reported back, so readiness fails and endpoints are no longer backed. That's what we should not take as an example, but rather as anti-pattern, I believe.